### PR TITLE
Drift Server

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/HeaderCodec.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        if (message instanceof ThriftFrame) {
+            ThriftFrame thriftFrame = (ThriftFrame) message;
+            verify(thriftFrame.getSequenceId().isPresent(), "Sequence id not set in response frame");
+            ByteBuf frame = encodeFrame(new HeaderFrame(
+                    thriftFrame.getSequenceId().getAsInt(),
+                    thriftFrame.getMessage(),
+                    thriftFrame.getHeaders(),
+                    HeaderTransportProtocol.valueOf(thriftFrame.getProtocolFactory()),
+                    thriftFrame.isSupportOutOfOrderResponse()));
+            context.write(frame, promise);
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/HeaderCodec.java]: How about
```java
if (message instanceof ThriftFrame) {
    ...
    message = encodeFrame(...);
}
context.write(message, promise);

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/SimpleFrameCodec.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext context, Object message)
+            throws Exception
+    {
+        if (message instanceof ByteBuf) {
+            ByteBuf request = (ByteBuf) message;
+            if (request.isReadable()) {
+                context.fireChannelRead(new ThriftFrame(OptionalInt.empty(), request, ImmutableMap.of(), protocolFactory, assumeClientsSupportOutOfOrderResponses));
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/SimpleFrameCodec.java]: Wrap arguments

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftFrame.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+{
+    private final OptionalInt sequenceId;
+    private final ByteBuf message;
+    private final Map<String, String> headers;
+    private final TProtocolFactory protocolFactory;
+    private final boolean supportOutOfOrderResponse;
+
+    public ThriftFrame(OptionalInt sequenceId, ByteBuf message, Map<String, String> headers, TProtocolFactory protocolFactory, boolean supportOutOfOrderResponse)
+    {
+        this.sequenceId = sequenceId;
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftFrame.java]: Check for null

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/OptionalSslHandler.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    public OptionalSslHandler(SslContext sslContext)
+    {
+        this.sslContext = sslContext;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out)
+            throws Exception
+    {
+        // 5 bytes are required to detect ssl
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/OptionalSslHandler.java]:
```java
// minimum bytes required to detect ssl

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        this.maxFrameSize = requireNonNull(maxFrameSize, "maxFrameSize is null");
+        this.thriftServerHandler = requireNonNull(thriftServerHandler, "thriftServerHandler is null");
+        this.assumeClientsSupportOutOfOrderResponses = assumeClientsSupportOutOfOrderResponses;
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out)
+            throws Exception
+    {
+        // 4 bytes are required to detect framing
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java]:
```java
// minimum bytes required to detect framing

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+    private void switchToFramedTransport(ChannelHandlerContext context, TProtocolFactory protocolFactory)
+    {
+        ChannelPipeline pipeline = context.pipeline();
+        new LengthPrefixedMessageFraming(maxFrameSize).addFrameHandlers(pipeline);
+        pipeline.addLast(new SimpleFrameCodec(protocolFactory, assumeClientsSupportOutOfOrderResponses));
+        pipeline.addLast(new ResponseOrderingHandler());
+        pipeline.addLast(thriftServerHandler);
+
+        // remove(this) must be last because it triggers down stream processing of the current message
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java]: "down stream" -> downstream

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+    private void switchToUnframedTransport(ChannelHandlerContext context, TProtocolFactory protocolFactory)
+    {
+        ChannelPipeline pipeline = context.pipeline();
+        new NoMessageFraming(protocolFactory, maxFrameSize).addFrameHandlers(pipeline);
+        pipeline.addLast(new SimpleFrameCodec(protocolFactory, assumeClientsSupportOutOfOrderResponses));
+        pipeline.addLast(new ResponseOrderingHandler());
+        pipeline.addLast(thriftServerHandler);
+
+        // remove(this) must be last because it triggers down stream processing of the current message
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java]: "down stream" -> downstream

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+    private void switchToHeaderTransport(ChannelHandlerContext context)
+    {
+        ChannelPipeline pipeline = context.pipeline();
+        new LengthPrefixedMessageFraming(maxFrameSize).addFrameHandlers(pipeline);
+        pipeline.addLast(new HeaderCodec());
+        pipeline.addLast(new ResponseOrderingHandler());
+        pipeline.addLast(thriftServerHandler);
+
+        // remove(this) must be last because it triggers down stream processing of the current message
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftProtocolDetection.java]: "down stream" -> downstream

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+                }
+
+                @Override
+                public void onFailure(Throwable t)
+                {
+                    context.disconnect();
+                }
+            });
+        }
+        catch (Exception e) {
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Does this happen only for bugs? Or can we get here if the client sends garbage data?

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            boolean supportOutOfOrderResponse)
+            throws Exception
+    {
+        long start = System.nanoTime();
+        TProtocol protocol = protocolFactory.getProtocol(new TChannelBufferInputTransport(frame));
+
+        TMessage message = protocol.readMessageBegin();
+        MethodMetadata method = methodInvoker.getMethodMetadata(message.getName());
+        if (method == null) {
+            return Futures.immediateFuture(writeApplicationException(
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Static import

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+        // set defaults for missing arguments
+        for (ParameterMetadata parameter : method.getParameters()) {
+            if (arguments[parameter.getIndex()] == null) {
+                Type argumentType = parameter.getCodec().getType().getJavaType();
+
+                if (argumentType instanceof Class) {
+                    Class<?> argumentClass = (Class<?>) argumentType;
+                    argumentClass = Primitives.unwrap(argumentClass);
+                    arguments[parameter.getIndex()] = Defaults.defaultValue(argumentClass);
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Static import

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+                }
+            }
+        }
+
+        return Arrays.asList(arguments);
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Static import

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+                    exception);
+
+            return new ThriftFrame(
+                    OptionalInt.of(sequenceId),
+                    buffer,
+                    ImmutableMap.of(),
+                    protocolFactory,
+                    supportOutOfOrderResponse);
+        }
+        else {
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: else is redundant

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+
+        TChannelBufferOutputTransport transport = new TChannelBufferOutputTransport(context.alloc().buffer(1024));
+        TProtocol protocol = protocolFactory.getProtocol(transport);
+
+        protocol.writeMessageBegin(new TMessage(methodName, EXCEPTION, sequenceId));
+
+        ExceptionWriter.writeTApplicationException(applicationException, protocol);
+
+        protocol.writeMessageEnd();
+        return new ThriftFrame(OptionalInt.of(sequenceId), transport.getOutputBuffer(), ImmutableMap.of(), protocolFactory, supportOutOfOrderResponse);
```
Comment (electrum) [inline drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Wrap arguments

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/DriftNettyServerConfig.java [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        this.sslEnabled = sslEnabled;
+        return this;
+    }
+
+    public File getTrustCertificate()
+    {
+        return trustCertificate;
+    }
+
+    @Config("thrift.server.ssl.trust-certificate")
```
[the remark this answers is not in the document]
Comment (electrum) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/DriftNettyServerConfig.java]: We discussed and decided to leave it as-is.

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/DriftNettyServerConfig.java [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+        this.sessionTimeout = sessionTimeout;
+        return this;
+    }
+
+    public List<String> getCiphers()
+    {
+        return ciphers;
+    }
+
+    @Config("thrift.server.ssl.ciphers")
```
[the remark this answers is not in the document]
Comment (electrum) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/DriftNettyServerConfig.java]: Follow up on this. See how the HTTP server does it.

File: drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java (hunk 9 above)
Comment (electrum) [reply drift-transport-netty/src/main/java/io/airlift/drift/transport/netty/server/ThriftServerHandler.java]: Should this also close the connection on Throwable?

File: drift-server/src/main/java/io/airlift/drift/server/ServiceMethod.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+            this.method = requireNonNull(method, "method is null");
+        }
+
+        @Override
+        public ListenableFuture<Object> invoke(ServerInvokeRequest request)
+        {
+            try {
+                List<Object> parameters = request.getParameters();
+                Object response = method.invoke(service, parameters.toArray(new Object[parameters.size()]));
+                if (response instanceof ListenableFuture) {
```
[the remark this answers is not in the document]
Comment (dain) [reply drift-server/src/main/java/io/airlift/drift/server/ServiceMethod.java]: Offline we decided to not make this change for now

File: drift-server/src/main/java/io/airlift/drift/server/stats/MethodInvocationStat.java [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.airlift.drift.server.stats;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+public interface MethodInvocationStat
+{
+    void recordResult(long startTime, ListenableFuture<Object> result);
```
[the remark this answers is not in the document]
Comment (dain) [reply drift-server/src/main/java/io/airlift/drift/server/stats/MethodInvocationStat.java]: The problem is the api takes startTime for the request and the *future* result, so we can not calculate duration at this point.

**Pull request merged**